File: 907-koko-eating-bananas/koko-eating-bananas.py
class Solution:

    def getHours(self, piles: List[int], rate:int ):
        count:int = 0
        for pile in piles:
            count += ceil(pile/rate)
        return count

    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        low = 1
        high = max(piles)
        sol = -1
        
        while( low <= high ):
            mid = low + (high - low)//2
            hours = self.getHours(piles, mid)
            if hours > h:
                low = mid + 1
            else:
                sol = min(mid, sol) if sol != -1 else mid 
                high = mid - 1
        
        return sol 

File: 907-koko-eating-bananas/koko-eating-bananas.cpp
class Solution {
public:
    bool canFit(vector<int>& piles,int h, int val){
        if(val < 1) return false;
        long long cycles_required = 0;
        for(const auto& pile: piles){
            cycles_required += ceil((double)pile/ val);
        }
        return cycles_required <= h;
    }

    int minEatingSpeed(vector<int>& piles, int h) {
        int best_freq = -1;
        int min_freq = 1;
        int max_freq = *max_element(piles.begin(), piles.end());
        while(min_freq <= max_freq){
            int mid = min_freq + (max_freq - min_freq)/2;
            if(canFit(piles, h, mid)){
                best_freq = mid;
                max_freq = mid - 1;
            } else {
                min_freq = mid + 1;
            }
        }
        return best_freq;
    }
};